framework/src/araui/painter/ui_painter.c

```diff
@@ -0,0 +1,602 @@
+{
+	ui_hal_draw_pixel(center_x + x, center_y + y, color);
+	ui_hal_draw_pixel(center_x + y, center_y + x, color);
+	ui_hal_draw_pixel(center_x - y, center_y + x, color);
+	ui_hal_draw_pixel(center_x - x, center_y + y, color);
+	ui_hal_draw_pixel(center_x - x, center_y - y, color);
+	ui_hal_draw_pixel(center_x - y, center_y - x, color);
+	ui_hal_draw_pixel(center_x + y, center_y - x, color);
+	ui_hal_draw_pixel(center_x + x, center_y - y, color);
+}
+
+ui_error_t ui_painter_draw_circle(int32_t center_x, int32_t center_y, int32_t radius, ui_color_t color)
+{
+	int32_t x;
+	int32_t y;
+	int epsilon;
+
+	if (radius < 0) {
+		return UI_INVALID_PARAM;
+	}
+
+	x = 0;
+	y = radius;
+	epsilon = 5 - 4 * radius;
+
+	_ui_painter_draw_circle_pixel(center_x, center_y, x, y, color);
+
+	while (y >= x) {
+		x++;
+		if (epsilon > 0) {
+			y--;
+			epsilon = epsilon + 8 * (x - y) + 4;
+		} else {
+			epsilon = epsilon + 8 * x + 4;
+		}
+		_ui_painter_draw_circle_pixel(center_x, center_y, x, y, color);
+	}
+
+	return UI_OK;
+}
+
+static void _ui_painter_draw_ellipse_pixel(int32_t center_x, int32_t center_y, int32_t x, int32_t y, ui_color_t color)
+{
+	ui_hal_draw_pixel(center_x + x, center_y + y, color);
+	ui_hal_draw_pixel(center_x - x, center_y + y, color);
+	ui_hal_draw_pixel(center_x + x, center_y - y, color);
+	ui_hal_draw_pixel(center_x - x, center_y - y, color);
+}
+
+ui_error_t ui_painter_draw_ellipse(int32_t center_x, int32_t center_y, int32_t radius_x, int32_t radius_y, ui_color_t color)
+{
+	int32_t x;
+	int32_t y;
+	int32_t radi_x_sq;
+	int32_t radi_y_sq;
+	int epsilon;
+
+	if (radius_x < 0 || radius_y < 0) {
+		return UI_INVALID_PARAM;
+	}
+
+	x = 0;
+	y = radius_y;
+	radi_x_sq = radius_x * radius_x;
+	radi_y_sq = radius_y * radius_y;
+	epsilon = 4 * radi_y_sq - 4 * radi_x_sq * radius_y + 1;
+
+	_ui_painter_draw_ellipse_pixel(center_x, center_y, x, y, color);
+
+	while ((radi_y_sq * x < radi_x_sq * y) && (x < radius_x)) {
+		x++;
+		if (epsilon >= 0) {
+			y--;
+			epsilon = epsilon + radi_y_sq * (2 * x + 3) + radi_x_sq * (3 - 2 * y);
+		} else {
+			epsilon = epsilon + radi_y_sq * (2 * x + 3);
+		}
+		_ui_painter_draw_ellipse_pixel(center_x, center_y, x, y, color);
+	}
+
+	while (y > 0 && y < radius_y) {
+		y--;
+		if (epsilon <= 0) {
+			x++;
+			epsilon = epsilon + radi_x_sq * (3 - 2 * y) + radi_y_sq * (2 * x + 3);
+		} else {
+			epsilon = epsilon + radi_x_sq * (3 - 2 * y);
+		}
+		_ui_painter_draw_ellipse_pixel(center_x, center_y, x, y, color);
+	}
+
+	return UI_OK;
+}
+
+static void _ui_painter_draw_arc(int32_t center_x, int32_t center_y,
+	int32_t start_x, int32_t start_y, int32_t end_x, int32_t end_y,
+	int32_t radius, int32_t quadrant, ui_color_t color)
+{
+	int32_t x;
+	int32_t y;
+	int epsilon;
+
+	if (quadrant < 1 || quadrant > 4) {
+		UI_LOGE("error: Invalid parameter!\n");
+		return;
+	}
+
+	x = start_x;
+	y = start_y;
+
+	ui_hal_draw_pixel(center_x + x, center_y + y, color);
+
+	if ((start_x == end_x) && (start_y == end_y)) {
```

Comment:
You don't have to check this condition at the start of function?

Reply:
At least one pixel should be drew.